app/src/main/java/com/futurice/freesound/feature/audio/ExoPlayerAudioPlayer.java

```diff
@@ -147,28 +157,46 @@ private void handleToggleAction(@NonNull final ToggleAction action,
 
     }
 
-    private void play(@NonNull final String url) {
-        exoPlayer.prepare(mediaSourceFactory.create(get(url)));
+    private void play(@NonNull final PlaybackSource playbackSource) {
+        exoPlayer.prepare(mediaSourceFactory.create(get(playbackSource.url())));
         exoPlayer.setPlayWhenReady(true);
-        currentUrl.set(Option.ofObj(url));
+        currentSource.set(Option.ofObj(playbackSource));
     }
 
     private void toggle(@NonNull final ToggleAction action) {
         exoPlayer.setPlayWhenReady(action != ToggleAction.PAUSE);
     }
 
     private void stop() {
-        currentUrl.set(Option.none());
+        currentSource.set(Option.none());
         exoPlayer.stop();
     }
 
     private static boolean isIdle(final int state) {
         return state == ExoPlayer.STATE_IDLE || state == ExoPlayer.STATE_ENDED;
     }
 
-    private boolean hasSourceUrlChanged(@NonNull final String url) {
-        return currentUrl.get()
-                         .map(id -> !id.equals(url))
-                         .orDefault(() -> false);
+    private boolean hasSourceChanged(@NonNull final PlaybackSource source) {
+        return currentSource.get()
+                            .map(playbackSource -> !playbackSource.equals(source))
+                            .orDefault(() -> false);
+    }
+
+    @NonNull
+    private static PlayerState.State toState(@NonNull final ExoPlayerState exoPlayerState) {
+        int exoplaybackState = exoPlayerState.playbackState();
+        switch (exoplaybackState) {
+            case ExoPlayer.STATE_IDLE:
+                return PlayerState.State.IDLE;
+            case ExoPlayer.STATE_BUFFERING:
+                return PlayerState.State.BUFFERING;
+            case ExoPlayer.STATE_READY:
+                return exoPlayerState.playWhenReady() ?
```

Comment:
Do you have any particular preference here, because I don't usually have any strong convention for this.